Mathlib/Algebra/Star/CentroidHom.lean

```diff
@@ -0,0 +1,142 @@
+/-
+Copyright (c) 2024 Christopher Hoskin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christopher Hoskin
+-/
+
+import Mathlib.Algebra.Ring.CentroidHom
+import Mathlib.Algebra.Star.StarRingHom
+import Mathlib.Algebra.Star.Subsemiring
+import Mathlib.Algebra.Star.Basic
+
+/-!
+# Centroid homomorphisms on Star Rings
+
+When a (non unital, non-associative) semi-ring is equipped with an involutive automorphism the
+center of the centroid becomes a star ring in a natural way and the natural mapping of the centre of
+the semi-ring into the centre of the centroid becomes a *-homomorphism.
+
+## Tags
+
+centroid
+-/
+
+variable {α : Type*}
+
+namespace CentroidHom
+
+section NonUnitalNonAssocStarSemiring
+
+variable [NonUnitalNonAssocSemiring α] [StarRing α]
+
+instance : Star (CentroidHom α) where
+  star f :=
+  { toFun := fun a => star (f (star a))
+    map_zero' := by
+      simp only [star_zero, map_zero]
+    map_add' := fun a b => by simp only [star_add, map_add]
+    map_mul_left' := fun a b => by simp only [star_mul, map_mul_right, star_star]
+    map_mul_right' := fun a b => by simp only [star_mul, map_mul_left, star_star] }
+
+@[simp] lemma star_apply (f : CentroidHom α) (a : α) : (star f) a = star (f (star a)) := rfl
+
+instance instInvolutiveStar : InvolutiveStar (CentroidHom α) where
+  star_involutive _ := ext (fun _ => by
+    rw [star_apply, star_apply, star_star, star_star])
+
+instance instStarAddMonoid : StarAddMonoid (CentroidHom α) where
+  star_involutive f := ext (fun _ => by
+    rw [star_apply, star_apply, star_star, star_star])
+  star_add f g := ext fun _ => star_add _ _
+
+instance : Star (Subsemiring.center (CentroidHom α)) where
+  star f := ⟨star (f : CentroidHom α), Subsemiring.mem_center_iff.mpr (fun g => ext (fun a =>
```

Comment:
Does this follow more directly from `Set.star_mem_center`? I'm not at lean right now, but happy to push some cleanups myself.

I wonder if we should have an inherited star instance on all submonoids / subrings

Reply:
Sorry, I now see this comment is silly.